Call next only once when request finishes

NickOvt opened the pull request:
Fixes #

Changes proposed in this pull request:

- Sometimes next plugins would exhibit Cannot pipe while piping error when piping to message_stream, indicating that rspamd plugin incorrectly handles next() call. This PR tries to fix this issue by calling next() only when the request stream has actually closed, either due to an error or an end of the request.
- Fix mfaddr handling, remove libmime and clean input from broken bytes (soft clean without enforcing ASCII)

NickOvt committed:
doc(CONTRIBUTORS): updated

NickOvt committed:
Merge branch 'master' of https://github.com/haraka/haraka-plugin-rspamd

NickOvt committed:
git fetchMerge branch 'master' of https://github.com/NickOvt/haraka-plugin-rspamd

NickOvt committed:
nextOnce - call next when request finished. Mfaddr fixes

msimerson committed:
Merge branch 'master' into ZMSA-35

msimerson commented on index.js:
```diff
@@ -111,7 +111,18 @@ exports.get_options = function (connection) {
     const mfaddr = connection.transaction.mail_from.address().toString()
 
-    if (mfaddr) options.headers.From = mfaddr
+    // If cannot encode to utf-8 mime string then default to manual sanitanization
```
```suggestion
    // manual sanitanization
```

msimerson committed:
revert mfrom handling and comments

Removed manual sanitization of email addresses and adjusted comments for clarity.

msimerson committed:
remove braces

msimerson commented on index.js:
```diff
@@ -220,7 +220,16 @@ exports.hook_data_post = function (next, connection) {
+    // wait until request stream closes
+    req.once('close', () => {
+      next(code, msg)
+    })
```
Should I be concerned that adding this listener here, versus before the `.pipe(req)`, may be adding the listener after the close event fires? It seems a bit racy to me.

NickOvt replied:
To be honest this whole PR is work in progress as I'm thinking how to best handle this.
The thing is it is required to make sure req stream closes so that next hook can pipe to message_stream.
So I might update this PR tomorrow

msimerson commented on index.js:
```diff
@@ -220,7 +220,16 @@ exports.hook_data_post = function (next, connection) {
+      next(code, msg)
+    })
+
+    // if we need to go to next hook then destroy request
```
what does this comment mean? Who is we? Don't we *always* go to the next hook?

msimerson commented on index.js:
```diff
@@ -220,7 +220,16 @@ exports.hook_data_post = function (next, connection) {
+
+    // if we need to go to next hook then destroy request
+    if (!req.destroyed) {
+      req.destroy()
```
In normal operation, this shouldn't be needed, right? 

If you're going to call `req.destroy()` here, perhaps you should be emitting the optional error event so that an error gets logged when this does fire. My suspicion is that you need this only because your `req.once('close'...)` was added after the request close was already emitted.

NickOvt committed:
call next when message_stream ends, clean request stream

NickOvt committed:
merge changes from remote

NickOvt commented:
I've updated the PR. Now when calling nextOnce we clean the request stream and remove it for good measure, then we drain the message_stream and call next when the message_stream ends.

NickOvt commented:
@msimerson

msimerson commented on index.js:
```diff
@@ -214,13 +214,26 @@ exports.hook_data_post = function (next, connection) {
   function nextOnce(code, msg) {
+    if (req) {
+      req.destroy()
+    }
```
style: please collapse this to a single line

msimerson commented on index.js:
```diff
@@ -214,13 +214,26 @@ exports.hook_data_post = function (next, connection) {
   let timer
   const timeout = plugin.cfg.main.timeout || plugin.timeout - 1
 
+  let req
```
is this change required?  (I expect not, due to variable hoisting).

NickOvt commented:
Hello! 

This PR is not relevant anymore and thus I think it is better to close it.
Cheers!
